python/unitytrainers/self_play_trainer_controller.py

```diff
@@ -67,41 +70,62 @@ def create_variable_mapping(self, from_scope, to_scope):
         :return: Dictionary mapping string variable names of scope :param from_scope: to the tf.Variable
         of scope :parm to_scop:
         """
-        variables_from_scope = tf.get_collection(tf.GraphKeys.GLOBAL_VARAIBLES, scope=from_scope)
-        variables_to_scope   = tf.get_collection(tf.GraphKeys.GLOBAL_VARAIBLES, scope=to_scope)
+        variables_from_scope = tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES, scope=from_scope)
+        variables_to_scope   = tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES, scope=to_scope)
         mapping = dict(zip(map(lambda x: x.name[:-2], variables_from_scope), variables_to_scope))
         return mapping
 
     def save_models_and_checkpoints(self, sess, global_step, global_saver):
         super(SelfPlayTrainerController, self).save_models_and_checkpoints(sess, global_step, global_saver)
-        if global_step % self.save_freq == 0 and global_step != 0 and self.train_model:
-            self.logger.debug("Saving new opponents behaviour")
+        if global_step % self.ghost_save_frequency == 0 and global_step != 0 and self.train_model:
             checkpoint_path = self.ghost_trainers_path + 'main_brain' + '-' + str(global_step) + '.cptk'
             self.main_trainer_saver.save(sess, checkpoint_path)
+            print("Saving new opponents behaviour. Total checkpoints: {}".format(self.main_brain_saved_checkpoints))
+            self.main_brain_saved_checkpoints.append(checkpoint_path)
 
     def handle_episode_termination(self, curr_info, sess):
         # handle opponent policy change
-        super(SelfPlayTrainerController, self).handle_episode_termination(curr_info, sess)
-        self.elapsed_episodes += 1
-        if self.should_change_ghost_model():
-            self.resample_all_ghosts(sess)
+        if self.env.global_done:
+            self.elapsed_episodes += 1
+            if self.should_change_ghost_model():
+                self.resample_all_ghosts(sess)
+        return super(SelfPlayTrainerController, self).handle_episode_termination(curr_info, sess)
 
     def resample_all_ghosts(self, sess):
+        """
+        Samples a historical policy for all ghost trainers
+        :param sess: Tensorflow session.
+        """
+        self.store_main_latest_policy(sess, self.main_trainer_saver, policy_path=self.latest_policy_checkpoint_path)
```

Comment:
Do we need `policy_path=`? Isn't this implicit from the argument orer?

Reply:
I wrote it purely for readability purposes, the interpreter can infer what it is by argument order. I would rather keep it, if it's all the same

Reply:
I don't feel strongly about it but I'm not sure what's particular about this argument that makes you want to use it here and not in other locations.

Reply:
Would you prefer if we also explicitly stated the parameter names? I actually feel that for this highly technical academic code it is useful to be explicit about the operations being made.